src/DocumentDB.ChangeFeedProcessor/LeaseManagement/DocumentServiceLease.cs

```diff
@@ -32,9 +35,26 @@ public DocumentServiceLease(DocumentServiceLease other)
             this.Properties = other.Properties;
         }
 
-        [JsonProperty("id")]
+        [JsonProperty(IdPropertyName)]
         public string Id { get; set; }
 
+        /// <summary>
+        /// Gets or sets property to be used as partition key path for lease collections created in GremlinAccounts
+        /// </summary>
+        [JsonProperty(GremlinCompatIdPropertyName)]
+        public string GremlinCompatId
```

Comment:
We should not do that by default. Majority of lease collections do not need this overhead.

Reply:
Can you suggest, what has to be done. I tried to interpret this comment below but not sure whether I got it right?

I can change gremlincompatId into LeaseId property name so that its not specific to gremlin and hopefully LeaseId doesnt impact any of other types of cosmosdb account. if you have a way where we can change Id into LeaseId thats ideal considering already collections are provisioned with partitionKey as "Id".

When you are saying its overhead, what's the concern. Is it processing cost, storage cost? clone of Id field is not costly right.
Are you saying we should create a derived class DocumentServiceLeaseV2 and add this property LeaseId there. That would be major refactoring to have parallel classes or if else within existing classes e.g DocumentServiceLeaseStoreManager.cs to conditionally create DocumentServiceLease vs DocumentServiceLeaseV2 based on user input. Overhead cost of additional string is cheaper to this entire effort.

Reply:
One way I think about is using something like (existing) IRequestOptionsFactory, that would use something like Lease.AlternatePartitionKey and all operations would append PK to request options for lease operations. That would only be used when CFP Builder explicitly opts-in for using alternate PK for lease collection.

Reply:
is it something along this line, I will fix the names later
internal class PartitionedByAlternatePKCollectionRequestOptionsFactory : IRequestOptionsFactory
    {
        public RequestOptions CreateRequestOptions(ILease lease) => new RequestOptions { PartitionKey = new PartitionKey(lease.AlternatePK) };

        public FeedOptions CreateFeedOptions() => new FeedOptions { EnableCrossPartitionQuery = true };
    }

then instantiate that class in CFBuilder based on some flag.

wont this mean the following:

1. ILease interface has to be updated with a property name "AlternatePK"
2. if somebody implemented their custom lease management, they need to unnecessarily add this property. Is It Ok?
3. the moment I touch ILease, i need to update DocumentServiceLease to have that property. so whether I use new RequestOptionsFactory, lease document will still contain the "AlternatePK". so what does it help in?
4.  if somebody created a lease collection with "FooId" as partitionkey. Still not clear, how I can make it work without touching ILease interface to have "FooId" as a property name. Or is this not a ask at all from you and Matias?

Reply:
ILease doesn't need to be touched. DocumentServiceLease is only specific to CosmosDB service, e.g. redis or in memory lease implementation wouldn't have partition key concept. I think with these changes we can add properties to document service lease either by adding to the class itself or by calling SetProerty via NewtonsoftJson as you did in your ealier changes. I would prefer the latter approach with use controlling what PK they want to use (in the upper stack we would call API to get PK from the collection).

Reply:
can you check this pr - https://github.com/Azure/azure-documentdb-changefeedprocessor-dotnet/pull/159/files
and see if that aligns with what you are proposing
this is rough change and ignore the test cases, I haven't fixed or authored new ones yet